orttraining/orttraining/core/optimizer/qdq_fusion.cc

```diff
@@ -0,0 +1,129 @@
+                                               const InlinedHashSet<std::string_view>& compatible_execution_providers) {
+  // Try to match the current node with QuantizeLinear in the effort of searching for the pattern
+  // QuantizeLinear -> DequqnatizeLinear.
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(quantize_node, "QuantizeLinear", {10, 13}) ||
+      !graph_utils::IsSupportedProvider(quantize_node, compatible_execution_providers)) {
+    return {false, nullptr};
+  }
+
+  // QuantizeLinear node found. Assert that there is only one node consuming the output of the QuantizeLinear node.
+  ORT_ENFORCE(quantize_node.GetOutputEdgesCount() == 1, "There must only be a single node consuming the output of ",
+              "QuantizeLinear node. Actual: The output of QuantizeLinear (", quantize_node.Name(),
+              ") is being consumed by ", quantize_node.GetOutputEdgesCount(), " nodes.");
+
+  // Now try to match the succeeding node with DequantizeLinear to ascertain QuantizeLinear -> DequantizeLinear
+  // pattern.
+  Node* dequantize_node_ptr = graph.GetNode(quantize_node.OutputNodesBegin()->Index());
+
+  // Every Q must be followed by a single and unique DQ and every DQ must be preceded by a single and unique Q.
+  // And hence every Q->DQ pair can be fused into a single and unique FakeQuant.
+  // So, we error out and let the user know that the graph is not what is expected if DequantizeLinear is not found.
+  ORT_ENFORCE(dequantize_node_ptr, "Expected: A DequantizeLinear node, Actual: nullptr.");
+  ORT_ENFORCE(graph_utils::IsSupportedOptypeVersionAndDomain(*dequantize_node_ptr, "DequantizeLinear", {10, 13}) &&
+                  graph_utils::IsSupportedProvider(*dequantize_node_ptr, compatible_execution_providers),
+              "Expected that every QuantizeLinear node be followed by a unique DequantizeLinear node. ",
+              "Actual: QuantizeLinear (", quantize_node.Name(), ") is followed by ", dequantize_node_ptr->OpType(), "(",
+              dequantize_node_ptr->Name(), ").");
+
+  return {true, dequantize_node_ptr};
+}
+
+}  // namespace
+
+Status QDQFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& nodes_in_topological_order = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : nodes_in_topological_order) {
+    auto* quantize_node_ptr = graph.GetNode(node_index);
+    if (!quantize_node_ptr) continue;  // Node was removed.
+
+    auto& quantize_node = *quantize_node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(quantize_node, modified, graph_level, logger));
+
+    // Check for QDQ pattern
+    auto [qdq_pattern_found, dequantize_node_ptr] =
+        CheckForQDQPatternMatch(graph, quantize_node, GetCompatibleExecutionProviders());
+    if (!qdq_pattern_found) {
+      continue;
+    }
+
+    // QuantizeLinear zero_point is INT8 or UINT8. FakeQuant uses quant_zero_point as FLOAT.
```

Comment:
Can you explain this : "FakeQuant uses quant_zero_point as FLOAT." I dont understand. Thanks!

Reply:
In a quantized onnx model, the zero point can be represented as either uint8 or int8 type.
The primary reason I am using the zero point as a float in qat is to support future scenarios where we might want to learn it during backpropagation. 

I could potentially change the implementation such that the zero point input can be any of int8, uint8 or float and the cuda kernel will be specialized for these three types. But this would not provide any additional benefit as such since all computation would still happen in float and would lead to (a small) increase in binary size. The only benefit would be that the zp type would be kept the same as it was in the quantized onnx model.

We can sync up to discuss this further.